core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -429,10 +434,19 @@ public FetchSearchResult executeFetchPhase(ShardFetchRequest request, SearchTask
             throw ExceptionsHelper.convertToRuntime(e);
         } finally {
             cleanContext(context);
+            //Release searchContext if we have no more hits
+            if ((context.queryResult().size()) >= context.scrollContext().totalHits) {
+                if (context.scrollContext() != null) {
+                    releasedContexts.add(context.id());
+                    freeContext(context.id());
+                }
+            }
         }
     }
 
     private SearchContext findContext(long id) throws SearchContextMissingException {
```

Comment:
can you add some docs saying that this may return null if the context was already fully consumed?